public/app/features/explore/state/epics.ts

```diff
@@ -0,0 +1,125 @@
+}
+
+export const startSubscriptionAction = actionCreatorFactory<StartSubscriptionPayload>(
+  'explore/START_SUBSCRIPTION'
+).create();
+
+export interface SubscriptionDataReceivedPayload {
+  data: SeriesData;
+  exploreId: ExploreId;
+}
+
+export const subscriptionDataReceivedAction = actionCreatorFactory<SubscriptionDataReceivedPayload>(
+  'explore/SUBSCRIPTION_DATA_RECEIVED'
+).create();
+
+export const startSubscriptionsEpic: Epic<ActionOf<any>, ActionOf<any>, StoreState> = (action$, state$) => {
+  return action$.ofType(startSubscriptionsAction.type).pipe(
+    mergeMap((action: ActionOf<StartSubscriptionsPayload>) => {
+      const { exploreId, dataReceivedActionCreator } = action.payload;
+      const { datasourceInstance, queries, refreshInterval } = state$.value.explore[exploreId];
+
+      if (!datasourceInstance || !datasourceInstance.convertToStreamTargets) {
+        return EMPTY; //do nothing if datasource does not support streaming
+      }
+
+      if (!refreshInterval || !isLive(refreshInterval)) {
+        return EMPTY; //do nothing if refresh interval is not 'LIVE'
+      }
+
+      const request: any = { targets: queries };
+      return datasourceInstance.convertToStreamTargets(request).map(target =>
+        startSubscriptionAction({
+          url: convertToWebSocketUrl(target.url),
+          refId: target.refId,
+          exploreId,
+          dataReceivedActionCreator,
+        })
+      );
+    })
+  );
+};
+
+export const startSubscriptionEpic: Epic<ActionOf<any>, ActionOf<any>, StoreState> = (action$, state$) => {
+  return action$.ofType(startSubscriptionAction.type).pipe(
+    mergeMap((action: ActionOf<StartSubscriptionPayload>) => {
+      const { url, exploreId, refId, dataReceivedActionCreator } = action.payload;
+      return webSocket(url).pipe(
+        takeUntil(
+          action$
+            .ofType(
+              startSubscriptionAction.type,
+              resetExploreAction.type,
+              updateDatasourceInstanceAction.type,
+              changeRefreshIntervalAction.type
+            )
+            .pipe(
+              filter(action => {
+                if (action.type === resetExploreAction.type) {
+                  return true; // stops all subscriptions if user navigates away
+                }
+
+                if (action.type === updateDatasourceInstanceAction.type && action.payload.exploreId === exploreId) {
+                  return true; // stops subscriptions if user changes data source
+                }
+
+                if (action.type === changeRefreshIntervalAction.type && action.payload.exploreId === exploreId) {
+                  return !isLive(action.payload.refreshInterval); // stops subscriptions if user changes refresh interval away from 'Live'
+                }
+
+                return action.payload.exploreId === exploreId && action.payload.refId === refId;
+              }),
+              tap(value => console.log('Stopping subscription', value))
```

Comment:
Hmm, you're right usually you send something like 'unsubscribe' over the line. I'll have to chat with the Loki team. The important thing is that any closed subscription doesn't receive any new data because that will cause very hard bugs to solve.

Reply:
> The important thing is that any closed subscription doesn't receive any new data because that will cause very hard bugs to solve.

Yeah sorry forget to mention that important thing - it works as you would expect :+1:

Reply:
I would prefer if the websocket handling was inside the data source and that the live updates was pushed via the existing DataStreamObserver https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/types/datasource.ts#L240  callback. That way we do not need resultToSeries data here and we can use the existing streaming interface. The epic here would still be needed to handle unsubscribe logic. 

So when initiating live mode explore would issue a query with a live flag on the query and then loki data source could create a DataStreamState and push updates via that.

Reply:
> I would prefer if the websocket handling was inside the data source and that the live updates was pushed via the existing DataStreamObserver https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/types/datasource.ts#L240 callback. That way we do not need resultToSeries data here and we can use the existing streaming interface. The epic here would still be needed to handle unsubscribe logic.
> 
> So when initiating live mode explore would issue a query with a live flag on the query and then loki data source could create a DataStreamState and push updates via that.

I'm not sure I understand how the DataStreamObserver solution would flesh out, I think I need to learn a lot more about that whole concept before I attempt to refactor.

Reply:
It’s a callback that contains a stream state & update , see testdata datasource it has steam handlers that use it. Think we need to use it if we ever want Loki streaming in a dashboard table or logs panel. Can try to walk through the code on next Monday frontend weekly

Reply:
Thanks, the walk through the code would be awesome!